api/apps/api/src/modules/scenarios/planning-units/scenario-planning-units-linker-service.ts

```diff
@@ -0,0 +1,121 @@
+export class ScenarioPlanningUnitsLinkerService {
+  constructor(
+    @InjectRepository(
+      ScenariosPlanningUnitGeoEntity,
+      DbConnections.geoprocessingDB,
+    )
+    private readonly puRepo: Repository<ScenariosPlanningUnitGeoEntity>,
+    @InjectRepository(Project)
+    private readonly projectsRepo: Repository<Project>,
+  ) {}
+
+  /** Currently we only support linking regular planning unit geometries
+   * computed via ST_HexagonGrid or ST_SquareGrid */
+  private isPlanningUnitGridShapeRegular(
+    shape: PlanningUnitGridShape | undefined,
+  ): boolean {
+    return !isNil(shape) && [
+      PlanningUnitGridShape.hexagon,
+      PlanningUnitGridShape.square,
+    ].includes(shape);
+  }
+
+  private isProjectUsingCustomPlanningArea(project: Project): boolean {
+    return !isNil(project.planningAreaGeometryId) && isNil(project.countryId) && isNil(project.adminAreaLevel1Id) && isNil(project.adminAreaLevel2Id);
+  }
+
+  private isProjectUsingGadmPlanningArea(project: Project): boolean {
+    return isNil(project.planningAreaGeometryId) && !(isNil(project.countryId) || isNil(project.adminAreaLevel1Id) || isNil(project.adminAreaLevel2Id));
+  }
+
+  private isProjectUsingRegularPlanningUnitGrid(project: Project): boolean {
+    return this.isPlanningUnitGridShapeRegular(project.planningUnitGridShape) && !isNil(project.planningUnitAreakm2);
+  }
+
+  private isProjectUsingCustomPlanningUnitGrid(project: Project): boolean {
+    return project.planningUnitGridShape === PlanningUnitGridShape.fromShapefile && isNil(project.planningUnitAreakm2);
+  }
+
+  private getQueryPartsForLinker(project: Project): QueryPartsForLinker {
+    /**
+     * @TODO selection by planning_units_geom.project_id needs project_id column
+     * to be added and set.
+     *
+     * We still intersect in case planning unit grid is not fully included in
+     * the planning area.
+     */
+    if(this.isProjectUsingCustomPlanningArea(project) && this.isProjectUsingCustomPlanningUnitGrid(project)) {
+      return {
+        planningUnitIntersectionQueryPart: `type = '${PlanningUnitGridShape.fromShapefile}' and project_id = '${project.id}'`,
+        planningUnitSelectionQueryPart: `(select the_geom from planning_areas where project_id = ${project.id}`,
+      }
+    }
+
+    if(this.isProjectUsingCustomPlanningArea(project) && this.isProjectUsingRegularPlanningUnitGrid(project)) {
+      return {
+        planningUnitIntersectionQueryPart: `type = '${PlanningUnitGridShape.fromShapefile}' and project_id = '${project.id}'`,
+        planningUnitSelectionQueryPart: `type = '${project.planningUnitGridShape}' and size = ${project.planningUnitAreakm2}`,
+      }
+    }
+
+    if(this.isProjectUsingGadmPlanningArea(project) && this.isProjectUsingCustomPlanningUnitGrid(project)) {
+      const adminAreaId = project.adminAreaLevel2Id ?? project.adminAreaLevel1Id ?? project.countryId;
+      const adminAreaLevel = AdminAreasService.levelFromId(adminAreaId);
```

Comment:
if we are only using some static method (util in this case), importing whole module shouldn't be necessary.

Reply:
do you mean to avoid static methods and promoting utility functions to module exports? e.g.

```
export const levelFromId = (areaId: string): number => {
  return areaId.match(/\./g)?.length ?? 0;
}
```

and

```
import { levelFromId } from AdminAreasService;
```

Reply:
if so, I'd actually quite favour this in general, since the (ES) module already works as a namespace so having the function as a class method adds namespacing redundancy, in practice.

Reply:
@hotzevzl in this particular case - yes, some util could be enough. 

What I meant is that you can still use this static method without importing the module / injecting service (as it doesn't use the DI and any of the module's services instances directly).

Although, using just a util seems a fair trade for me, as creating connection (even if not through DI) to different module is a step forward to "everything uses everything" :)